### PR TITLE
feat: generate samples

### PR DESCRIPTION
**Pull request by JustinBeckwith**
This PR removes the code that downloads sample snippets from GCS, and replaces it with an in-house generator.  For reasons I don't fully understand, dartman was only generating samples for a small select number of APIs.  It was easy enough to make this generic, so I rolled it out broadly.  

*The interesting files are at the bottom of the list, so scroll down*

TODO:
- [x] Generate one sample per API method in standalone files
- [x] Get look-behind regex working for projects/my-project paths

After this lands:
- [ ] Update whatever tool cloudsite uses to inline current GCS / disco-artifact-generator samples
- [ ] Delete code in disco-artifact manager that produces nodejs samples

Stuff to think about:
- I dropped the generation of pagination code.  If we care that much, I'll add a pagination feature.
- The plan is to eventually generate standalone samples as well to replace the dartman generated ones used on cloudsite.  All the bits are here, lets just save it for a followup PR.

### CODE REVIEW

File: src/generator/generator.ts [hunk 1]
```diff
@@ -11,7 +11,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-import * as fs from 'fs';
+import {promises as fs} from 'fs';
```
Comment (alexander-fenster) [inline src/generator/generator.ts:14]: so... this one is gnarly, it's not allowed with `engines >= 10`, needs 10.something. Maybe stick to `util.promisify` for some more time?

File: src/generator/generator.ts [hunk 2]
```diff
@@ -69,6 +68,33 @@ function isSimpleType(type: string): boolean {
...
+  // remove trailing $
+  if (pattern.endsWith('$')) {
+    pattern = pattern.slice(0, pattern.length - 1);
+  }
+  // replace projects placeholders
+  pattern = pattern.replace('projects/[^/]+', 'projects/my-project');
+  pattern = pattern.replace('locations/[^/]+', 'locations/my-location');
+  // TODO: This should be replaced with a proper regular express
+  // that does a look behind to find the name to be replaced.  Alex
+  // will probably know how to do it.
```
Comment (alexander-fenster) [inline src/generator/generator.ts:94]: Oh, you don't need look-behind here. You only need look-behinds when the string being looked-behind is unknown. Here, it's easier:

```
> pattern = "^projects/[^/]+/locations/[^/]+/products/[^/]+/referenceImages/[^/]+$"
'^projects/[^/]+/locations/[^/]+/products/[^/]+/referenceImages/[^/]+$'
> pattern.replace(/\^?(\w+)s\/\[\^\/\]\+\$?/g, "$1s/my-$1")
'projects/my-project/locations/my-location/products/my-product/referenceImages/my-referenceImage'
```

You only need to make sure that different regex applications do not overlap, other than that, it's pretty straightforward, with some added backslashes. We just hope they have all plurals ending with "s", I can't singularize in regex properly :)

```
description:   optional ^     project  s   /  [   ^   /   ]  +  optional $
      regex: /     \^?        (\w+)    s  \/ \[  \^  \/  \] \+     \$?      /
```
and then we replace it with `$1s/my-$1` which expands to `projects/my-project`.

File: src/generator/templates/sample.njk [hunk 3]
```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 Google LLC
```
Comment (alexander-fenster) [inline src/generator/templates/sample.njk:1]: Don't we have a license in a separate included `.njk` as we do in gapic-generator-typescript?
```njk
{% import "_license.njk" as license -%}
{% license.license() -%}
```

File: src/generator/templates/sample.njk [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  // Acquire an auth client, and bind it to all future calls
+  const authClient = await auth.getClient();
+  google.options('auth', authClient);
+
+  // Do the magic
+  const res = await {{method.id}}({
+    {% if method.parameters %}
+    {% for pname, p in method.parameters|dictsort %}
+    {% if p.description %} // {{ p.description|oneLine|cleanComments|safe }}{% endif %}
```
Comment (alexander-fenster) [inline src/generator/templates/sample.njk:51]: I don't really like `oneLine`, I see nothing bad if comments spread into multiple lines, like we do in `jsdoc`.

File: src/generator/templates/sample.njk [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    
+    {{ pname|cleanPropertyName|safe }}: {% if p.pattern %}'{{ p.pattern|unRegex }}'{% else %}'placeholder-value'{% endif %},
+    {% endfor %}
+    {% endif %}
+    {% if method.request %}
+
+    // Request body metadata
+    requestBody: {
+      // request body parameters
```
Comment (alexander-fenster) [inline src/generator/templates/sample.njk:60]: Yeah, I guess we need to expand this as well, do we?

File: src/generator/templates/sample.njk (hunk 3 above)
Comment (JustinBeckwith) [reply src/generator/templates/sample.njk:1]: For you #2057

File: src/generator/templates/sample.njk (hunk 4 above)
Comment (JustinBeckwith) [reply src/generator/templates/sample.njk:51]: Tend to agree.  Fixed!

File: src/generator/templates/sample.njk (hunk 5 above)
Comment (JustinBeckwith) [reply src/generator/templates/sample.njk:60]: What do you mean?

File: src/generator/generator.ts (hunk 2 above)
Comment (JustinBeckwith) [reply src/generator/generator.ts]: This worked like a dream.  Thank you!

File: src/generator/templates/sample.njk (hunk 5 above)
Comment (alexander-fenster) [reply src/generator/templates/sample.njk:60]: I mean, does `requestBody` need to contain some API-specific fields that should be listed in the sample?

Comment (JustinBeckwith) [reply src/generator/templates/sample.njk:60]: We can totally do that.  Turns out it's just *really* annoying :)

File: src/generator/generator.ts (hunk 1 above)
Comment (JustinBeckwith) [reply src/generator/generator.ts:14]: Only really needed to run the generator, so I'm not too concerned :)

**Comment by bcoe** [comment]
> I dropped the generation of pagination code. If we care that much, I'll add a pagination feature.

I wonder if we could show an async iterator in action, they simplify pagination a ton.

**Comment by JustinBeckwith** [comment]
@alexander-fenster @bcoe this is ready for a look for real

File: src/apis/abusiveexperiencereport/v1.ts [hunk 6]
```diff
@@ -167,6 +167,55 @@ export namespace abusiveexperiencereport_v1 {
     /**
      * abusiveexperiencereport.sites.get
      * @desc Gets a site's Abusive Experience Report summary.
+     * @example
+     * // Before running the sample:
+     * // - Enable the API at:
+     * //   https://console.developers.google.com/apis/api/abusiveexperiencereport
```
Comment (alexander-fenster) [inline src/apis/abusiveexperiencereport/v1.ts:173]: This link did not work for me. Not sure if it's something with my account, or the link is wrong. Just FYI.

File: src/apis/abusiveexperiencereport/v1.ts [hunk 7]
```diff
@@ -167,6 +167,55 @@ export namespace abusiveexperiencereport_v1 {
...
+     *   });
+     *
+     *   // Acquire an auth client, and bind it to all future calls
+     *   const authClient = await auth.getClient();
+     *   google.options('auth', authClient);
+     *
+     *   // Do the magic
+     *   const res = await abusiveexperiencereport.sites.get({
+     *     // Required. The name of the site whose summary to get, e.g.
+     *     // `sites/http%3A%2F%2Fwww.google.com%2F`.
```
Comment (alexander-fenster) [inline src/apis/abusiveexperiencereport/v1.ts:195]: Is this supposed to be like this, or do we need `| safe`?  (I really don't know)

File: src/generator/filters.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+export function isSimpleType(type: string): boolean {
+  if (type.indexOf('{') > -1) {
+    return false;
+  }
+  return true;
+}
+
+export function cleanPropertyName(prop: string) {
+  const match = prop.match(/[-@.]/g);
```
Comment (alexander-fenster) [inline src/generator/filters.ts:40]: nit: `/g` is not needed here

File: src/generator/filters.ts [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+ * @param regex pattern for the given parameter
+ */
+export function unRegex(regex: string): string {
+  // example: ^projects/[^/]+$' ==> projects/my-project
+  let pattern = regex;
+  if (typeof regex !== 'string') {
+    return '';
+  }
+  // remove leading ^
+  if (pattern.startsWith('^')) {
```
Comment (alexander-fenster) [inline src/generator/filters.ts:160]: The mega-regex in line 168 should take care of leading `^` and trailing `$`.

File: src/generator/generator.ts [hunk 10]
```diff
@@ -431,27 +224,17 @@ export class Generator {
...
+    const apiPath = path.join(srcPath, 'apis', schema.name);
+    const exportFilename = path.join(apiPath, schema.version + '.ts');
     await mkdirp(path.dirname(exportFilename));
+    // populate the `method.fragment` property with samples
+    addFragments(schema);
+    // generate the API (ex: src/apis/youtube/v3.ts)
     await this.render(API_TEMPLATE, {api: schema}, exportFilename);
+    // generate samples on disk at:
+    // src/apis/<service>/samples/<version>/<method>
+    // generateSamples(apiPath, schema);
```
Comment (alexander-fenster) [inline src/generator/generator.ts:237]: commented out on purpose?

File: src/apis/abusiveexperiencereport/v1.ts (hunk 6 above)
Comment (JustinBeckwith) [reply src/apis/abusiveexperiencereport/v1.ts:173]: Ugh.  Updating this to append `.googleapis.com` (not needed in all of them)

File: src/apis/abusiveexperiencereport/v1.ts (hunk 7 above)
Comment (JustinBeckwith) [reply src/apis/abusiveexperiencereport/v1.ts:195]: So I *do* use `| safe`.  Turns out this is exactly what's in the source discovery doc 🤦

File: src/generator/generator.ts (hunk 10 above)
Comment (JustinBeckwith) [reply src/generator/generator.ts:237]: Yes!  The code works, but that's what write the samples to disk.  I don't want to do that during this first pass.

**Comment by JustinBeckwith** [comment]
@bcoe I'm entirely open to adding the pagination stuff with async iterators.  As a first step, we need to hand craft a sample with tests and such.  Any concerns with punting that to a followup PR?

File: src/apis/abusiveexperiencereport/v1.ts [hunk 11]
```diff
@@ -257,6 +306,43 @@ export namespace abusiveexperiencereport_v1 {
...
+     *
+     * async function main() {
+     *   const auth = new google.auth.GoogleAuth({
+     *     // Scopes can be specified either as an array or as a single, space-delimited string.
+     *     scopes: [],
+     *   });
+     *
+     *   // Acquire an auth client, and bind it to all future calls
+     *   const authClient = await auth.getClient();
+     *   google.options('auth', authClient);
```
Comment (bcoe) [inline src/apis/abusiveexperiencereport/v1.ts:329]: I like that there are now inline examples, we should double check that our refdoc generator renders these appropriately at deploy time.

File: src/generator/templates/sample.njk [hunk 12]
```diff
@@ -0,0 +1,77 @@
+{% if standalone %}
+{% include "license.njk" %}
+
+// THIS FILE IS AUTO-GENERATED
```
Comment (bcoe) [inline src/generator/templates/sample.njk:4]: what generates this file?

File: src/apis/abusiveexperiencereport/v1.ts (hunk 11 above)
Comment (JustinBeckwith) [reply src/apis/abusiveexperiencereport/v1.ts:329]: Fun fact: inline samples are generally broken in compodoc 😭 
https://github.com/compodoc/compodoc/issues/304

The inline example is no more broken than it was before I guess?  I think @fhinkel is working on a plan to save us

File: src/generator/templates/sample.njk (hunk 12 above)
Comment (JustinBeckwith) [reply src/generator/templates/sample.njk:4]: Hah, this is hand written - I want the files *it generates* to have that comment.

**Pull request merged**